packages/hoppscotch-js-sandbox/src/utils.ts

```diff
@@ -1,89 +1,439 @@
+import * as O from "fp-ts/Option"
+import { pipe } from "fp-ts/lib/function"
+import { cloneDeep } from "lodash-es"
 
-export function marshalObjectToVM(
-  vm: QuickJS.QuickJSVm,
-  obj: object
-): E.Either<string, QuickJS.QuickJSHandle> {
-  let jsonString
+import {
+  GlobalEnvItem,
+  SelectedEnvItem,
+  TestDescriptor,
+  TestResult,
+} from "./types"
 
-  try {
-    jsonString = JSON.stringify(obj)
-  } catch (e) {
-    return E.left("Marshaling stringification failed")
+const getEnv = (envName: string, envs: TestResult["envs"]) => {
+  return O.fromNullable(
+    envs.selected.find((x: SelectedEnvItem) => x.key === envName) ??
+      envs.global.find((x: GlobalEnvItem) => x.key === envName)
+  )
+}
+
+// Compiles shared scripting API methods for use in both pre and post request scripts
+const getSharedMethods = (envs: TestResult["envs"]) => {
+  let updatedEnvs = envs
+
+  const envGetFn = (key: any) => {
+    if (typeof key !== "string") {
+      throw new Error("Expected key to be a string")
+    }
+
+    const result = pipe(
+      getEnv(key, updatedEnvs),
+      O.match(
+        () => undefined,
+        ({ value }) => String(value)
+      )
+    )
+
+    return result
   }
 
-  const vmStringHandle = vm.newString(jsonString)
+  const envGetResolveFn = (key: any) => {
+    if (typeof key !== "string") {
+      throw new Error("Expected key to be a string")
+    }
 
-  const jsonHandle = vm.getProp(vm.global, "JSON")
-  const parseFuncHandle = vm.getProp(jsonHandle, "parse")
+    const result = pipe(
+      getEnv(key, updatedEnvs),
+      E.fromOption(() => "INVALID_KEY" as const),
 
-  const parseResultHandle = vm.callFunction(
-    parseFuncHandle,
-    vm.undefined,
-    vmStringHandle
-  )
+      E.map(({ value }) =>
+        pipe(
+          parseTemplateStringE(value, [
+            ...updatedEnvs.selected,
+            ...updatedEnvs.global,
+          ]),
+          // If the recursive resolution failed, return the unresolved value
+          E.getOrElse(() => value)
+        )
+      ),
+      E.map((x) => String(x)),
 
-  if (parseResultHandle.error) {
-    parseResultHandle.error.dispose()
-    return E.left("Marshaling failed")
+      E.getOrElseW(() => undefined)
+    )
+
+    return result
   }
 
-  const resultHandle = vm.unwrapResult(parseResultHandle)
+  const envSetFn = (key: any, value: any) => {
+    if (typeof key !== "string") {
+      throw new Error("Expected key to be a string")
+    }
 
-  vmStringHandle.dispose()
-  parseFuncHandle.dispose()
-  jsonHandle.dispose()
+    if (typeof value !== "string") {
+      throw new Error("Expected value to be a string")
+    }
 
-  return E.right(resultHandle)
-}
+    updatedEnvs = setEnv(key, value, updatedEnvs)
 
-export function getEnv(envName: string, envs: TestResult["envs"]) {
-  return O.fromNullable(
-    envs.selected.find((x) => x.key === envName) ??
-      envs.global.find((x) => x.key === envName)
-  )
+    return undefined
+  }
```

Comment:
As mentioned earlier, when this method is invoked, there isn't a way to return the updated environment entries to the surrounding context. The return value below from the outside function gets processed beforehand. Hence, we rely on mutation. This only applies to `utils`; we obtain a deep copy of the arg supplied from the outside world.